internal/charmstore/store.go

```diff
@@ -493,19 +525,109 @@ func (s *Store) FindBestEntity(url *charm.URL, fields map[string]int) (*mongodoc
 			"promulgated-revision": 1,
 			"series":               1,
 			"revision":             1,
+			"development":          1,
+			"stable":               1,
 		}
 		for f := range fields {
 			nfields[f] = 1
 		}
 		fields = nfields
 	}
+	if url.Revision != -1 {
+		// If the URL contains a revision, then it refers to a single entity.
+		entity, err := s.findSingleEntity(url, fields)
+		if errgo.Cause(err) == params.ErrNotFound {
+			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+		} else if err != nil {
+			return nil, errgo.Mask(err)
+		}
+		// If a channel was spedified make sure the entity is in that channel.
+		if channel == StableChannel && !entity.Stable {
+			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+		}
+		if channel == DevelopmentChannel && !entity.Development {
+			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+		}
+		return entity, nil
+	}
+	switch channel {
+	case StableChannel, UnspecifiedChannel:
+		return s.findStableEntity(url, fields)
+	case DevelopmentChannel:
+		return s.findDevelopmentEntity(url, fields)
+	case UnpublishedChannel:
+		return s.findUnpublishedEntity(url, fields)
+	default:
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	}
+}
+
+// findStableEntity attempts to find an entity on the stable channel. The
+// base entity for URL is retrieved and the series with the best match to
+// URL.Series is used as the resolved entity.
+func (s *Store) findStableEntity(url *charm.URL, fields map[string]int) (*mongodoc.Entity, error) {
+	baseEntity, err := s.FindBaseEntity(url, map[string]int{
+		"_id":          1,
+		"stableseries": 1,
+	})
+	if errgo.Cause(err) == params.ErrNotFound {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	} else if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	entityURL := baseEntity.StableSeries[url.Series]
+	if entityURL == nil && url.Series == "" {
+		for _, u := range baseEntity.StableSeries {
+			if entityURL == nil || seriesScore[u.Series] > seriesScore[entityURL.Series] {
+				u := u
+				entityURL = u
+			}
+		}
+	}
+	if entityURL == nil {
+		return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
+	}
+	return s.findSingleEntity(entityURL, fields)
+}
+
+// findDevelopmentEntity attempts to find an entity on the development
+// channel. The base entity for URL is retrieved and the series with the
+// best match to URL.Series is used as the resolved entity.
+func (s *Store) findDevelopmentEntity(url *charm.URL, fields map[string]int) (*mongodoc.Entity, error) {
```

Comment:
When @rogpeppe's refactor happens it will be easier to remove the duplication, so I'll wait.